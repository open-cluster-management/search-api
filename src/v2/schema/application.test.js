/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2020. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project

import supertest from 'supertest';
import server, { GRAPHQL_PATH } from '../index';

describe('Application Resolver', () => {
  test('Correctly Resolves Application Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          applications {
            _uid
            clusterCount
            created
            dashboard
            labels
            name
            namespace
            hubSubscriptions {
              _uid
              timeWindow
              localPlacement
              status
              channel
            }
            hubChannels
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Single Application', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          applications (namespace: "test", name: "app02") {
            _uid
            clusterCount
            created
            dashboard
            labels
            name
            namespace
            hubSubscriptions {
              _uid
              status
              channel
            }
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Ignores filters when only name or namespace is passed.', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          applications (name: "app02") {
            _uid
            name
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
});

describe('Placement Rule Resolver', () => {
  test('Correctly Resolves Placement Rules Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          placementRules {
            _uid
            name
            namespace
            created
            selfLink
            clusterCount
            replicas
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Single Placement Rule', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          placementRules(namespace: "applications", name: "pr01") {
            _uid
            name
            namespace
            created
            selfLink
            clusterCount
            replicas
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Ignores filters when only name or namespace is passed.', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          placementRules (name: "sub02") {
            _uid
            name
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
});

describe('Subscription Resolver', () => {
  test('Correctly Resolves Subscriptions Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          subscriptions {
            _uid
            name
            namespace
            created
            selfLink
            timeWindow
            localPlacement
            status
            channel
            appCount
            clusterCount
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Single Subscription', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
      {
        subscriptions(namespace: "test", name: "sub02") {
          _uid
          name
          namespace
          created
          selfLink
          timeWindow
          localPlacement
          status
          channel
          appCount
          clusterCount
        }
      }
    `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Ignores filters when only name or namespace is passed.', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
      {
        subscriptions (name: "sub02") {
          _uid
          name
        }
      }
    `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
});

describe('Channel Resolver', () => {
  test('Correctly Resolves Channels Query', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          channels {
            _uid
            name
            namespace
            created
            selfLink
            type
            pathname
            localPlacement
            subscriptionCount
            clusterCount
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Correctly Resolves Single Channel', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          channels(namespace: "test", name: "ch02") {
            _uid
            name
            namespace
            created
            selfLink
            type
            pathname
            localPlacement
            subscriptionCount
            clusterCount
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));

  test('Ignores filters when only name or namespace is passed.', () => new Promise((done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        {
          channels (name: "ch02") {
            _uid
            name
          }
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  }));
});
