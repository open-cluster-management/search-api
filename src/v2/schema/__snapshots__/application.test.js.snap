// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application Resolver Correctly Resolves Application Query 1`] = `
Object {
  "data": Object {
    "applications": Array [
      Object {
        "_uid": "local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99",
        "clusterCount": Object {
          "localCount": 1,
          "remoteCount": 2,
        },
        "created": "2020-01-23T13:56:32Z",
        "dashboard": "localhost/grafana/dashboard/db/app01-dashboard-via-federated-prometheus?namespace=sample",
        "hubChannels": Array [
          Object {
            "app._uid": "local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99",
            "ch._uid": "local-cluster/233dfffd-f421-44ee-811b-7f3352b2d728",
            "ch.pathname": "https://github.com/fxiang1/app-samples.git",
            "ch.type": "Git",
            "sub._gitbranch": "master",
            "sub._gitcommit": "d67d8e10dcfa41dddcac14952e9872e1dfece06f",
            "sub._gitpath": "helloworld",
            "sub._uid": "local-cluster/bdced01f-3bd4-11ea-a488-00000a100f99",
            "sub.name": "sub1",
            "sub.namespace": "default",
          },
          Object {
            "app._uid": "local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99",
            "ch._uid": "local-cluster/6c8dcb97-5e6e-4606-9b55-ae3eb05fcfb5",
            "ch.pathname": "sample-ns",
            "ch.type": "Namespace",
            "sub._uid": "local-cluster/b218636d-3d5e-11ea-8ed1-00000a100f99",
            "sub.name": "sub2",
            "sub.namespace": "default",
          },
        ],
        "hubSubscriptions": Array [
          Object {
            "_uid": "local-cluster/bdced01f-3bd4-11ea-a488-00000a100f99",
            "channel": "dev1/dev1",
            "localPlacement": true,
            "status": null,
            "timeWindow": null,
          },
          Object {
            "_uid": "local-cluster/b218636d-3d5e-11ea-8ed1-00000a100f99",
            "channel": "default/mortgage-channel",
            "localPlacement": false,
            "status": "Propagated",
            "timeWindow": "blocked",
          },
          Object {
            "_uid": "local-cluster/66426f24-3bd3-11ea-a488-00000a100f99",
            "channel": "dev1/dev1",
            "localPlacement": false,
            "status": "Propagated",
            "timeWindow": "active",
          },
        ],
        "labels": Array [
          "label1=value1",
          "label2=value2",
        ],
        "name": "app01",
        "namespace": "sample",
        "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/sample/applications/app01",
      },
      Object {
        "_uid": "local-cluster/app-02-uid",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 0,
        },
        "created": "2020-01-23T13:56:32Z",
        "dashboard": "localhost/grafana/dashboard/db/app02-dashboard-via-federated-prometheus?namespace=test",
        "hubChannels": Array [
          Object {
            "app._uid": "local-cluster/app-02-uid",
            "ch._uid": "local-cluster/233dfffd-f421-44ee-811b-7f3352b2d728",
            "ch.pathname": "http://multiclusterhub-repo.open-cluster-management.svc.cluster.local:3000/charts",
            "ch.type": "HelmRepo",
            "sub._uid": "local-cluster/66426f24-3bd3-11ea-a488-00000a100f99",
            "sub.name": "sub3",
            "sub.namespace": "default",
            "sub.package": "application-chart",
            "sub.packageFilterVersion": "2.2.0",
          },
        ],
        "hubSubscriptions": Array [],
        "labels": Array [],
        "name": "app02",
        "namespace": "test",
        "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/test/applications/app02",
      },
    ],
  },
}
`;

exports[`Application Resolver Correctly Resolves Single Application 1`] = `
Object {
  "data": Object {
    "applications": Array [
      Object {
        "_uid": "local-cluster/app-02-uid",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 0,
        },
        "created": "2020-01-23T13:56:32Z",
        "dashboard": "localhost/grafana/dashboard/db/app02-dashboard-via-federated-prometheus?namespace=test",
        "hubSubscriptions": Array [],
        "labels": Array [],
        "name": "app02",
        "namespace": "test",
        "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/test/applications/app02",
      },
    ],
  },
}
`;

exports[`Application Resolver Ignores filters when only name or namespace is passed. 1`] = `
Object {
  "data": Object {
    "applications": Array [
      Object {
        "_uid": "local-cluster/29a848d6-3de8-11ea-9f0f-00000a100f99",
        "name": "app01",
      },
      Object {
        "_uid": "local-cluster/app-02-uid",
        "name": "app02",
      },
    ],
  },
}
`;

exports[`Application Resolver Ignores filters when only name or namespace is passed. 2`] = `
Object {
  "data": Object {
    "applications": Array [
      Object {
        "_uid": "local-cluster/app-02-uid",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 0,
        },
        "created": "2020-01-23T13:56:32Z",
        "dashboard": "localhost/grafana/dashboard/db/app02-dashboard-via-federated-prometheus?namespace=test",
        "hubSubscriptions": Array [],
        "labels": Array [],
        "name": "app02",
        "namespace": "test",
        "selfLink": "/apis/app.k8s.io/v1beta1/namespaces/test/applications/app02",
      },
    ],
  },
}
`;

exports[`Channel Resolver Correctly Resolves Channels Query 1`] = `
Object {
  "data": Object {
    "channels": Array [
      Object {
        "_uid": "local-cluster/ch-01-uid",
        "clusterCount": Object {
          "localCount": 1,
          "remoteCount": 2,
        },
        "created": "2020-08-20T14:16:05Z",
        "localPlacement": true,
        "name": "ch01",
        "namespace": "applications",
        "pathname": "http://multiclusterhub-repo.open-cluster-management.svc.cluster.local:3000/charts",
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/applications/channels/ch01",
        "subscriptionCount": 0,
        "type": "HelmRepo",
      },
      Object {
        "_uid": "local-cluster/ch-02-uid",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 2,
        },
        "created": "2020-08-20T14:17:05Z",
        "localPlacement": false,
        "name": "ch02",
        "namespace": "test",
        "pathname": "https://github.com/fxiang1/app-samples.git",
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/test/channels/ch02",
        "subscriptionCount": 3,
        "type": "Git",
      },
    ],
  },
}
`;

exports[`Channel Resolver Correctly Resolves Single Channel 1`] = `
Object {
  "data": Object {
    "channels": Array [
      Object {
        "_uid": "local-cluster/ch-02-uid",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 2,
        },
        "created": "2020-08-20T14:17:05Z",
        "localPlacement": false,
        "name": "ch02",
        "namespace": "test",
        "pathname": "https://github.com/fxiang1/app-samples.git",
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/test/channels/ch02",
        "subscriptionCount": 3,
        "type": "Git",
      },
    ],
  },
}
`;

exports[`Channel Resolver Ignores filters when only name or namespace is passed. 1`] = `
Object {
  "data": Object {
    "channels": Array [
      Object {
        "_uid": "local-cluster/ch-01-uid",
        "name": "ch01",
      },
      Object {
        "_uid": "local-cluster/ch-02-uid",
        "name": "ch02",
      },
    ],
  },
}
`;

exports[`Placement Rule Resolver Correctly Resolves Placement Rules Query 1`] = `
Object {
  "data": Object {
    "placementRules": Array [
      Object {
        "_uid": "local-cluster/pr-01-uid",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 0,
        },
        "created": "2020-08-20T14:16:05Z",
        "name": "pr01",
        "namespace": "applications",
        "replicas": 3,
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/placementrules/pr01",
      },
      Object {
        "_uid": "local-cluster/pr-02-uid",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 0,
        },
        "created": "2020-08-20T14:17:05Z",
        "name": "pr02",
        "namespace": "test",
        "replicas": null,
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/test/pr02",
      },
    ],
  },
}
`;

exports[`Placement Rule Resolver Correctly Resolves Single Placement Rule 1`] = `
Object {
  "data": Object {
    "placementRules": Array [
      Object {
        "_uid": "local-cluster/pr-01-uid",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 0,
        },
        "created": "2020-08-20T14:16:05Z",
        "name": "pr01",
        "namespace": "applications",
        "replicas": 3,
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/placementrules/pr01",
      },
    ],
  },
}
`;

exports[`Placement Rule Resolver Ignores filters when only name or namespace is passed. 1`] = `
Object {
  "data": Object {
    "placementRules": Array [
      Object {
        "_uid": "local-cluster/pr-01-uid",
        "name": "pr01",
      },
      Object {
        "_uid": "local-cluster/pr-02-uid",
        "name": "pr02",
      },
    ],
  },
}
`;

exports[`Subscription Resolver Correctly Resolves Single Subscription 1`] = `
Object {
  "data": Object {
    "subscriptions": Array [
      Object {
        "_uid": "local-cluster/sub-02-uid",
        "appCount": 2,
        "channel": "object-ch-ns/object-ch",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 2,
        },
        "created": "2020-08-20T14:17:05Z",
        "localPlacement": false,
        "name": "sub02",
        "namespace": "test",
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/test/sub02",
        "status": "PropagationFailed",
        "timeWindow": null,
      },
    ],
  },
}
`;

exports[`Subscription Resolver Correctly Resolves Subscriptions Query 1`] = `
Object {
  "data": Object {
    "subscriptions": Array [
      Object {
        "_uid": "local-cluster/sub-01-uid",
        "appCount": 1,
        "channel": "git-ch-ns/git-ch",
        "clusterCount": Object {
          "localCount": 1,
          "remoteCount": 2,
        },
        "created": "2020-08-20T14:16:05Z",
        "localPlacement": true,
        "name": "sub01",
        "namespace": "applications",
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/applications/sub01",
        "status": "Propagated",
        "timeWindow": "blocked",
      },
      Object {
        "_uid": "local-cluster/sub-02-uid",
        "appCount": 2,
        "channel": "object-ch-ns/object-ch",
        "clusterCount": Object {
          "localCount": 0,
          "remoteCount": 2,
        },
        "created": "2020-08-20T14:17:05Z",
        "localPlacement": false,
        "name": "sub02",
        "namespace": "test",
        "selfLink": "/apis/apps.open-cluster-management.io/v1/namespaces/test/sub02",
        "status": "PropagationFailed",
        "timeWindow": null,
      },
    ],
  },
}
`;

exports[`Subscription Resolver Ignores filters when only name or namespace is passed. 1`] = `
Object {
  "data": Object {
    "subscriptions": Array [
      Object {
        "_uid": "local-cluster/sub-01-uid",
        "name": "sub01",
      },
      Object {
        "_uid": "local-cluster/sub-02-uid",
        "name": "sub02",
      },
    ],
  },
}
`;
